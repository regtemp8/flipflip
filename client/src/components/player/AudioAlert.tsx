import React, {
  type PropsWithChildren,
  useEffect,
  useRef,
  useState,
  CSSProperties,
  useCallback
} from 'react'
import {
  ForwardedProps,
  UseTransitionResult,
  animated,
  useTransition
} from 'react-spring'

import { type Theme, Typography } from '@mui/material'
import { makeStyles } from 'tss-react/mui'

import { grey } from '@mui/material/colors'
import {
  selectAudioName,
  selectAudioThumb,
  selectAudioAlbum,
  selectAudioArtist,
  selectAudioUrl
} from '../../store/audio/selectors'
import { selectUndefined } from '../../store/app/selectors'
import { useAppSelector } from '../../store/hooks'

const useStyles = makeStyles()((theme: Theme) => ({
  alert: {
    float: 'left',
    margin: theme.spacing(5),
    display: 'flex'
  },
  thumb: {
    maxHeight: 250
  },
  infoContainer: {
    display: 'flex',
    position: 'relative',
    flexDirection: 'column-reverse'
  },
  infoBackdrop: {
    position: 'absolute',
    backgroundColor: grey[500],
    opacity: 0.5,
    filter: 'blur(5px)',
    width: '100%',
    height: '100%',
    zIndex: 8
  },
  info: {
    paddingLeft: theme.spacing(4),
    paddingRight: theme.spacing(4),
    zIndex: 9,
    '&:nth-child(2)': {
      paddingBottom: theme.spacing(4),
      textDecoration: 'underline'
    },
    '&:last-child': {
      paddingTop: theme.spacing(4)
    }
  }
}))

interface AudioAlertLayerProps {
  visible: boolean
}

function AudioAlertLayer(props: PropsWithChildren<AudioAlertLayerProps>) {
  const fadeTransitions: UseTransitionResult<
    boolean,
    ForwardedProps<CSSProperties>
  >[] = useTransition(props.visible, (visible: any) => visible, {
    from: {
      // Base values, optional
      opacity: 0
    },
    enter: {
      // Styles apply for entering elements
      opacity: 1
    },
    leave: {
      // Styles apply for leaving elements
      opacity: 0
    },
    config: {
      duration: 2000
    }
  })

  return (
    <React.Fragment>
      {fadeTransitions.map((transition) => {
        return (
          <animated.div
            key={transition.key}
            style={{
              zIndex: 10,
              position: 'absolute',
              bottom: 0,
              ...transition.props
            }}
          >
            {transition.item === true && props.children}
          </animated.div>
        )
      })}
    </React.Fragment>
  )
}

export interface AudioAlertProps {
  audioID?: number
}

function AudioAlert(props: AudioAlertProps) {
  const nameSelector =
    props.audioID != null ? selectAudioName(props.audioID) : selectUndefined
  const name = useAppSelector(nameSelector)
  const thumbSelector =
    props.audioID != null ? selectAudioThumb(props.audioID) : selectUndefined
  const thumb = useAppSelector(thumbSelector)
  const urlSelector =
    props.audioID != null ? selectAudioUrl(props.audioID) : selectUndefined
  const url = useAppSelector(urlSelector)
  const albumSelector =
    props.audioID != null ? selectAudioAlbum(props.audioID) : selectUndefined
  const album = useAppSelector(albumSelector)
  const artistSelector =
    props.audioID != null ? selectAudioArtist(props.audioID) : selectUndefined
  const artist = useAppSelector(artistSelector)

  const [visible, setVisible] = useState(false)
  const _timeout = useRef<number>()

  const show = useCallback(() => {
    setVisible(true)
    _timeout.current = window.setTimeout(hide, 6000)
  }, [])

  useEffect(() => {
    if (props.audioID != null) {
      show()
    }

    return () => {
      clearTimeout(_timeout.current)
      _timeout.current = undefined
    }
  }, [props.audioID, show])

  useEffect(() => {
    clearTimeout(_timeout.current)
    show()
  }, [props.audioID, show])

  const hide = () => {
    setVisible(false)
  }

  const { classes } = useStyles()
  if (props.audioID == null) return <React.Fragment />

  return (
    <AudioAlertLayer visible={visible}>
      <div className={classes.alert}>
        <img className={classes.thumb} src={thumb} alt={name} />
        <div className={classes.infoContainer}>
          <div className={classes.infoBackdrop} />
          <Typography variant="h3" className={classes.info}>
            {name ?? url}
          </Typography>
          {artist && (
            <Typography variant="h5" className={classes.info}>
              {artist}
            </Typography>
          )}
          {album && (
            <Typography variant="h6" className={classes.info}>
              {album}
            </Typography>
          )}
        </div>
      </div>
    </AudioAlertLayer>
  )
}

;(AudioAlert as any).displayName = 'AudioAlert'
export default AudioAlert
