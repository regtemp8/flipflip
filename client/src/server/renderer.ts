import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
