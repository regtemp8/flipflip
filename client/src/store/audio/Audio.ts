import { copy, TF } from 'flipflip-common'

export default interface Audio {
  [key: string]: number | number[] | string | boolean | undefined

  id: number
  url?: string
  marked: boolean
  tags: number[] // Array of Tag IDs
  volume: number
  speed: number
  stopAtEnd: boolean
  nextSceneAtEnd: boolean
  tick: boolean
  tickMode: string
  tickDelay: number
  tickMinDelay: number
  tickMaxDelay: number
  tickSinRate: number
  tickBPMMulti: number
  bpm: number
  thumb?: string
  name?: string
  artist?: string
  album?: string
  trackNum?: number
  duration?: number
  comment?: string
  playedCount: number
}

export const initialAudio: Audio = {
  id: 0,
  marked: false,
  tags: [],
  volume: 100,
  speed: 10,
  stopAtEnd: false,
  nextSceneAtEnd: false,
  tick: false,
  tickMode: TF.constant,
  tickDelay: 1000,
  tickMinDelay: 500,
  tickMaxDelay: 5000,
  tickSinRate: 100,
  tickBPMMulti: 10,
  bpm: 0,
  playedCount: 0
}

export function newAudio(init?: Partial<Audio>) {
  const audio = Object.assign(copy<Audio>(initialAudio), init)
  if (audio.tickMode === 'at.constant') {
    audio.tickMode = 'tf.c'
  } else if (audio.tickMode === 'at.random') {
    audio.tickMode = 'tf.random'
  } else if (audio.tickMode === 'at.sin') {
    audio.tickMode = 'tf.sin'
  } else if (audio.tickMode === 'at.scene') {
    audio.tickMode = 'tf.scene'
  }

  return audio
}
