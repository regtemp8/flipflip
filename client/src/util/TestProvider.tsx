import * as React from 'react'
import { Provider } from 'react-redux'
import { createTheme } from '@mui/material'
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles'
import defaultTheme from 'flipflip-common'

export default class TestProvider extends React.Component {
  readonly props: {
    children: any
    store: any
  }

  disableTransitions = {
    defaultProps: {
      disablePortal: true,
      hideBackdrop: true,
      TransitionComponent: ({ children }) => children
    }
  }

  render() {
    const theme = createTheme({
      ...defaultTheme,
      // disable Transitions when runing tests, throws errors otherwise.
      components: {
        MuiMenu: this.disableTransitions,
        MuiPopover: this.disableTransitions,
        MuiDialog: this.disableTransitions,
        MuiModal: this.disableTransitions
      }
    } as any)
    return (
      <Provider store={this.props.store}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    )
  }
}
