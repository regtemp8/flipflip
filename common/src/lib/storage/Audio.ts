import { TF } from '../const';
import { copy } from '../utils';

import { type Tag } from './Tag';

export type Audio = {
  id: number;
  url?: string;
  marked: boolean;
  tags: Tag[];
  volume: number;
  speed: number;
  stopAtEnd: boolean;
  nextSceneAtEnd: boolean;
  tick: boolean;
  tickMode: string;
  tickDelay: number;
  tickMinDelay: number;
  tickMaxDelay: number;
  tickSinRate: number;
  tickBPMMulti: number;
  bpm: number;
  thumb?: string;
  name?: string;
  artist?: string;
  album?: string;
  trackNum?: number;
  duration?: number;
  comment?: string;
  playedCount: number;
};

export const initialAudio: Audio = {
  id: 0,
  marked: false,
  tags: [],
  volume: 100,
  speed: 10,
  stopAtEnd: false,
  nextSceneAtEnd: false,
  tick: false,
  tickMode: TF.constant,
  tickDelay: 1000,
  tickMinDelay: 500,
  tickMaxDelay: 5000,
  tickSinRate: 100,
  tickBPMMulti: 10,
  bpm: 0,
  playedCount: 0,
};

export function newAudio(init?: Partial<Audio>): Audio {
  const tickModes: Record<string, string> = {};
  tickModes['at.constant'] = 'tf.c';
  tickModes['at.random'] = 'tf.random';
  tickModes['at.sin'] = 'tf.sin';
  tickModes['at.scene'] = 'tf.scene';

  const audio = Object.assign(copy<Audio>(initialAudio), init);
  const tickMode = tickModes[audio.tickMode];
  if (tickMode != null) {
    audio.tickMode = tickMode;
  }

  return audio;
}
