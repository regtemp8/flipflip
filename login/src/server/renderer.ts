import createCache, { type EmotionCache } from "@emotion/cache";

export function createEmotionCache(nonce?: string): EmotionCache {
  return createCache({ key: "css", nonce });
}
