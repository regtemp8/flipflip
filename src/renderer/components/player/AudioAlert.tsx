import React, {
  type PropsWithChildren,
  useEffect,
  useRef,
  useState
} from 'react'
import { animated, useTransition } from 'react-spring'

import { type Theme, Typography } from '@mui/material'
import createStyles from '@mui/styles/createStyles'
import withStyles, { type WithStyles } from '@mui/styles/withStyles'
import { grey } from '@mui/material/colors'

import type Audio from '../../../store/audio/Audio'

const styles = (theme: Theme) =>
  createStyles({
    alert: {
      float: 'left',
      margin: theme.spacing(5),
      display: 'flex'
    },
    thumb: {
      maxHeight: 250
    },
    infoContainer: {
      display: 'flex',
      position: 'relative',
      flexDirection: 'column-reverse'
    },
    infoBackdrop: {
      position: 'absolute',
      backgroundColor: grey[500],
      opacity: 0.5,
      filter: 'blur(5px)',
      width: '100%',
      height: '100%',
      zIndex: 8
    },
    info: {
      paddingLeft: theme.spacing(4),
      paddingRight: theme.spacing(4),
      zIndex: 9,
      '&:nth-child(2)': {
        paddingBottom: theme.spacing(4),
        textDecoration: 'underline'
      },
      '&:last-child': {
        paddingTop: theme.spacing(4)
      }
    }
  })

interface AudioAlertLayerProps {
  visible: boolean
}

function AudioAlertLayer (props: PropsWithChildren<AudioAlertLayerProps>) {
  const fadeTransitions: [{ item: any, props: any, key: any }] = useTransition(
    props.visible,
    (visible: any) => visible,
    {
      from: {
        // Base values, optional
        opacity: 0
      },
      enter: {
        // Styles apply for entering elements
        opacity: 1
      },
      leave: {
        // Styles apply for leaving elements
        opacity: 0
      },
      config: {
        duration: 2000
      }
    }
  )

  return (
    <React.Fragment>
      {fadeTransitions.map((transition) => {
        return (
          <animated.div
            key={transition.key}
            style={{
              zIndex: 10,
              position: 'absolute',
              bottom: 0,
              ...transition.props
            }}
          >
            {transition.item === true && props.children}
          </animated.div>
        )
      })}
    </React.Fragment>
  )
}

export interface AudioAlertProps extends WithStyles<typeof styles> {
  audio: Audio
}

function AudioAlert (props: AudioAlertProps) {
  const [visible, setVisible] = useState(false)

  const _timeout = useRef<number>()

  useEffect(() => {
    if (props.audio) {
      show()
    }

    return () => {
      clearTimeout(_timeout.current)
      _timeout.current = undefined
    }
  }, [])

  useEffect(() => {
    clearTimeout(_timeout.current)
    show()
  }, [props.audio])

  const show = () => {
    setVisible(true)
    _timeout.current = window.setTimeout(hide.bind(this), 6000)
  }

  const hide = () => {
    setVisible(false)
  }

  const classes = props.classes
  if (!props.audio) return <React.Fragment />

  return (
    <AudioAlertLayer visible={visible}>
      <div className={classes.alert}>
        <img className={classes.thumb} src={props.audio.thumb} />
        <div className={classes.infoContainer}>
          <div className={classes.infoBackdrop} />
          <Typography variant="h3" className={classes.info}>
            {props.audio.name ? props.audio.name : props.audio.url}
          </Typography>
          {props.audio.artist && (
            <Typography variant="h5" className={classes.info}>
              {props.audio.artist}
            </Typography>
          )}
          {props.audio.album && (
            <Typography variant="h6" className={classes.info}>
              {props.audio.album}
            </Typography>
          )}
        </div>
      </div>
    </AudioAlertLayer>
  )
}

;(AudioAlert as any).displayName = 'AudioAlert'
export default withStyles(styles)(AudioAlert as any)
