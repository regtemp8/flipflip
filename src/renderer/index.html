<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>FlipFlip</title>
    <!-- Responsive meta tag -- ensures proper rendering and touch zooming for all devices -->
    <meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no" />
    <!-- Roboto Font -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap" />
    <!-- Font Icons -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons" />

    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://www.imagefap.com http://www.imagefap.com https://hypno.nimja.com/ http://hypno.nimja.com/; font-src 'self' https://fonts.gstatic.com; style-src 'self' https://fonts.googleapis.com 'unsafe-inline'"
    />
  </head>

  <body>
    <div id="app" class="app"></div>
  </body>
</html>
