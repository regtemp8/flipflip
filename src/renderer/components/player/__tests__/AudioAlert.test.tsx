import React from "react";
import { describe, it, expect } from "@jest/globals";
import renderer from "react-test-renderer";
import AudioAlert from "../AudioAlert";
import TestProvider from "../../../../../test/util/TestProvider";
import Audio from "../../../data/Audio";

describe("AudioAlert", () => {
  it("should match snapshot", () => {
    const audio = new Audio();
    const component = renderer.create(
      <TestProvider>
        <AudioAlert audio={audio} />
      </TestProvider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
