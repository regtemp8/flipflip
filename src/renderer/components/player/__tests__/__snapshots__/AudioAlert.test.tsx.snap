// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioAlert should match snapshot 1`] = `
[
  <div
    style={
      {
        "bottom": 0,
        "opacity": 0,
        "position": "absolute",
        "zIndex": 10,
      }
    }
  />,
  <div
    style={
      {
        "bottom": 0,
        "opacity": 0,
        "position": "absolute",
        "zIndex": 10,
      }
    }
  >
    <div
      className="AudioAlert-alert-1"
    >
      <img
        className="AudioAlert-thumb-2"
      />
      <div
        className="AudioAlert-infoContainer-3"
      >
        <div
          className="AudioAlert-infoBackdrop-4"
        />
        <h3
          className="MuiTypography-root MuiTypography-h3 AudioAlert-info-5 css-gepadz-MuiTypography-root"
        />
      </div>
    </div>
  </div>,
]
`;
